pysnark/boolean.py

```diff
@@ -0,0 +1,186 @@
+
+"""
+Support for Boolean values
+Costs 0 constraints to convert a LinComb to a LinCombBool
+"""
+
+class LinCombBool:
+    def __init__(self, lc):
+        """
+        Constructs a LinCombBool
+        Costs 2 constraints
+        """
+        if not isinstance(lc, LinComb):
+            raise RuntimeError("Wrong type for LinCombBool")
+        if not LinCombBool.is_boolean_value(lc.value):
+            raise ValueError("LinCombBool can only take Boolean values")
+        
+        # Add boolean constraint to circuit
+        add_constraint(lc, 1 - lc, LinComb.ZERO)
+        
+        self.lc = lc
+
+    def val(self):
+        return self.lc.val()
+
+    def __repr__(self):
+        return "{" + str(self.lc.value) + "}"
+
+    @classmethod
+    def is_boolean_value(cls, val):
+        if val == 0 or val == 1 or val == True or val == False:
+            return True
+        return False
+
+    @classmethod
+    def parse_boolean(cls, val):
+        if val == 0 or val == 1:
+            return val
+        if val == True:
+            return 1
+        if val == False:
+            return 0
+        raise ValueError("LinCombBool can only take Boolean values")
+
+    @classmethod
+    def _ensurebool(cls, val):
+        if isinstance(val,LinCombBool):
+            return val
+        elif isinstance(val,LinComb):
+            if not LinCombBool.is_boolean_value(val.value):
+                raise ValueError("LinCombBool can only take Boolean values")
+            return LinCombBool(val)
+        elif isinstance(val, int):
+            if not LinCombBool.is_boolean_value(val):
+                raise ValueError("LinCombBool can only take Boolean values")
+            return LinCombBool(ConstVal(val))
+        raise RuntimeError("Wrong type for LinCombBool")
+    
+    def __add__(self, other):
+        return self.lc + other
+
+    def __sub__(self, other):
+        return self.lc - other
+
+    def __mul__(self, other):
+        return self.lc * (other)
+    
+    def __truediv__(self, other): 
+        return NotImplemented
+         
+    def __floordiv__(self, other):
+        return NotImplemented
+
+    def __mod__(self, other):
+        return NotImplemented
+        
+    def __divmod__(self, divisor):
+        return NotImplemented
+
+    __radd__ = __add__
+    __rmul__ = __mul__
+
+    def __rsub__(self, other):
+        return other + (-self.lc)
+
+    def __rtruediv__(self, other):
+        return NotImplemented
+
+    def __neg__(self):
+        """
+        Performs an arithmetic negation
+        Costs 0 constraints
+        """
+        return -self.lc
+    
+    def __invert__(self):
+        return ConstVal(1) - self.lc
+
+    def __and__(self, other):
+        """
+        Performs a logical AND
+        Costs 0 constraints to AND with a constant
+        Costs 1 constraint to AND with a LinComb
+        """
+        if isinstance(other, int):
+            return LinCombBool(self.lc * other)

```

Comment:
Another way to implement this would be

try:
  other = bool(other)
  return LinCombBool(self.lc*(1 if other else 0))
except:
  other = LinCombBool._ensurebool(other)
  return LinCombBool(self.lc*other.lc)

That way it wouldn't just work for ints but for other non-oblivious data types as well?

Reply:
Pushed the changes. Also found a bug in `LinCombBool.__or__` that I fixed.